src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,228 @@
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartDefaultAsync(CancellationToken ct = default)
+    {
+      // TODO: Use the State property?
+      if (defaultInstance != null && !defaultInstance.disposed)
+      {
+        return defaultInstance;
+      }
+
+      await DefaultLock.WaitAsync(ct)
+        .ConfigureAwait(false);
+
+      if (defaultInstance != null && !defaultInstance.disposed)
+      {
+        return defaultInstance;
+      }
+
+      try
+      {
+        defaultInstance = await GetAndStartNewAsync(DefaultSessionId, ct)
+          .ConfigureAwait(false);
+        return defaultInstance;
+      }
+      finally
+      {
+        DefaultLock.Release();
+      }
+    }
+
+    /// <summary>
+    /// Starts and returns a new <see cref="ResourceReaper" /> instance.
+    /// </summary>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static Task<ResourceReaper> GetAndStartNewAsync(CancellationToken ct = default)
+    {
+      return GetAndStartNewAsync(Guid.NewGuid(), ct);
+    }
+
+    /// <summary>
+    /// Starts and returns a new <see cref="ResourceReaper" /> instance.
+    /// </summary>
+    /// <param name="sessionId">The session id.</param>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartNewAsync(Guid sessionId, CancellationToken ct = default)
+    {
+      // TODO: Do we need to check if the session id is already in use?
+      var resourceReaper = new ResourceReaper(sessionId);
+
+      try
+      {
+        await resourceReaper.StartAsync(ct)
+          .ConfigureAwait(false);
+        return resourceReaper;
+      }
+      catch (Exception)
+      {
+        await resourceReaper.DisposeAsync()
+          .ConfigureAwait(false);
+        throw;
+      }
+    }
+
+    public ValueTask DisposeAsync()
+    {
+      if (this.disposed)
+      {
+        return default;
+      }
+
+      this.disposed = true;
+
+      this.cancellationTokenSource.Cancel();
+      this.cancellationTokenSource.Dispose();
+
+      return this.resourceReaperContainer?.DisposeAsync() ?? default;
+    }
+
+    private async Task StartAsync(CancellationToken ct = default)
+    {
+      var ryukInitializedTaskSource = new TaskCompletionSource<bool>();
+
+      await this.resourceReaperContainer.StartAsync(ct)
+        .ConfigureAwait(false);
+
+      _ = this.MaintainRyukConnection(ryukInitializedTaskSource, ct);
+
+      await ryukInitializedTaskSource.Task
+        .ConfigureAwait(false);
+    }
+
+    private async Task MaintainRyukConnection(TaskCompletionSource<bool> ryukInitializedTaskSource, CancellationToken ct = default)
```

Comment:
We can remove the default value for the `CancellationToken` as this private method is only called with this parameter.

Reply:
I removed the default value.